### PR TITLE
Add Altitude Hold to SFM+

### PR DESCRIPTION
**Pull request by BradMick**

### CODE REVIEW

**Review by mattysmith22 — CHANGES_REQUESTED** [review]

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/config/CfgUserActions.hpp [hunk 1]
```diff
@@ -33,5 +33,9 @@ class CfgUserActions
     CfgUserActionDef(fza_ah64_wasRkt, "WAS Rockets", "Action the rockets in your crew station");
     CfgUserActionDef(fza_ah64_wasMsl, "WAS Missiles", "Action the missiles in your crew station");
     CfgUserActionDef(fza_ah64_masterSafe, "Toggle Master Safe", "Arms or Safes the aircraft");
+    CfgUserActionDef(fza_ah64_forceTrimHoldModeSwitch_up, "Force Trim/Hold Mode Switch Up", "NOT CURRENTLY USED");
+	CfgUserActionDef(fza_ah64_forceTrimHoldModeSwitch_right, "Force Trim/Hold Mode Switch Right", "Activates Altitude Hold");
+	CfgUserActionDef(fza_ah64_forceTrimHoldModeSwitch_down, "Force Trim/Hold Mode Switch Down", "De-Activates both Attitude and Altitude Hold");
+	CfgUserActionDef(fza_ah64_forceTrimHoldModeSwitch_left, "Force Trim/Hold Mode Switch Left", "Activates Attitude Hold");
```
[the remark this answers is not in the document]
Comment (BradMick) [reply @AH-64D Apache Longbow/Addons/fza_ah64_controls/config/CfgUserActions.hpp:39]: Complete

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/config/UserActionGroups.hpp [hunk 2]
```diff
@@ -25,7 +25,11 @@ class UserActionGroups
             fza_ah64_wasGun,
             fza_ah64_wasRkt,
             fza_ah64_wasMsl,
-            fza_ah64_masterSafe
+            fza_ah64_masterSafe,
+			fza_ah64_forceTrimHoldModeSwitch_up,
```
[the remark this answers is not in the document]
Comment (BradMick) [reply @AH-64D Apache Longbow/Addons/fza_ah64_controls/config/UserActionGroups.hpp:29]: Complete

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/config/UserActionsConflictGroups.hpp [hunk 3]
```diff
@@ -19,7 +19,11 @@ class UserActionsConflictGroups{
             fza_ah64_wasGun,
             fza_ah64_wasRkt,
             fza_ah64_wasMsl,
-            fza_ah64_masterSafe
+            fza_ah64_masterSafe,
+			fza_ah64_forceTrimHoldModeSwitch_up,
```
[the remark this answers is not in the document]
Comment (BradMick) [reply @AH-64D Apache Longbow/Addons/fza_ah64_controls/config/UserActionsConflictGroups.hpp:23]: Complete

File: @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/event/fn_eventInit.sqf [hunk 4]
```diff
@@ -129,6 +129,15 @@ _heli setVariable ["fza_ah64_fire_apu_fx", []];
 _heli setVariable ["fza_ah64_audioQueList", []];
 
 [_heli] call fza_sfmplus_fnc_coreConfig;
+_heli setVariable ["fza_ah64_attHoldActive",     false,    true];
```
[the remark this answers is not in the document]
Comment (BradMick) [reply @AH-64D Apache Longbow/Addons/fza_ah64_controls/scripting/functions/event/fn_eventInit.sqf:132]: Moved

File: @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmc.sqf [hunk 5]
```diff
@@ -0,0 +1,10 @@
+params ["_heli", "_deltaTime"];
+
+private _attHoldCycPitchOut = 0.0;
+private _attHoldCycRollOut  = 0.0;
+
+private _hdgHoldPedalYawOut = 0.0;
+
+private _altHoldCollOut = [_heli, _deltaTime] call fza_sfmplus_fnc_fmcAltitudeHold;
```
[the remark this answers is not in the document]
Comment (BradMick) [reply @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmc.sqf:8]: Handled in next PR

File: @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAltitudeHold.sqf [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    [_pid] call fza_fnc_pidReset;
+};
+
+if ( _heli getVariable "fza_ah64_altHoldActive") then {
+    //If the pilot is intentionally trying to change altitude, de-activate altitude
+    //hold and allow them to do so
+    private _collRef_low = _collRef * 0.95;
+    private _collRef_hi  = _collRef * 1.05;
+    if (fza_sfmplus_collectiveOutput >= _collRef_hi || fza_sfmplus_collectiveOutput <= _collRef_low) then {
+        _heli setVariable ["fza_ah64_altHoldActive", false];
```
[the remark this answers is not in the document]
Comment (BradMick) [reply @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAltitudeHold.sqf:27]: Complete

File: @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAltitudeHold.sqf [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+    };
+
+    //If the helicopters radar altitude is < 1428ft (435.25m) and current velocity is < 40kts
+    //then set the desired altitude to the current AGL altitude, otherwise set it to the
+    //current ASL altitude.
+    if (_curAltAGL < 435.254 && _curVel < 20.577) then {
```
[the remark this answers is not in the document]
Comment (BradMick) [reply @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAltitudeHold.sqf:33]: Complete

File: @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/cfgFunctions.hpp [hunk 8]
```diff
@@ -29,6 +29,14 @@ class CfgFunctions
             class engineReset {R;};
 			class engineVariables {R;};
 		};
+        class fmc  {
+            file = "\fza_ah64_sfmplus\functions\fmc";
+			class fmc {R;};
+            class fmcAltitudeHold {R;};
+			class fmcAltitudeHoldEnable {R;};
+			class fmcAttitudeHoldEnable {R;};
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/cfgFunctions.hpp:37]: TAABS

File: @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAttitudeHoldEnable.sqf [hunk 9]
```diff
@@ -0,0 +1,7 @@
+params ["_heli"]
+
+if (_heli getVariable "fza_ah64_attHoldActive" == false) then {
+    _heli setVariable ["fza_ah64_attHoldActive", true];
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAttitudeHoldEnable.sqf:4]: Shouldn't these be public assignments?

File: @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAltitudeHoldEnable.sqf [hunk 10]
```diff
@@ -0,0 +1,33 @@
+params ["_heli"];
+
+if (_heli getVariable "fza_ah64_altHoldActive" == false) then {
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAltitudeHoldEnable.sqf:3]: Shouldn't these be public assignments?

File: @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcHoldModesDisable.sqf [hunk 11]
```diff
@@ -0,0 +1,9 @@
+params ["_heli"];
+
+//De-activate attitude hold and set the reference back to 0
+_heli setVariable ["fza_ah64_attHoldActive",     false];
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcHoldModesDisable.sqf:4]: Shouldn't these be public assignments?

**Review by mattysmith22 — CHANGES_REQUESTED** [review]

File: @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/cfgFunctions.hpp (hunk 8 above)
Comment (BradMick) [reply @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/cfgFunctions.hpp:37]: Fixed

File: @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAttitudeHoldEnable.sqf (hunk 9 above)
Comment (BradMick) [reply @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAttitudeHoldEnable.sqf:4]: Complete

File: @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcHoldModesDisable.sqf (hunk 11 above)
Comment (BradMick) [reply @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcHoldModesDisable.sqf:4]: Fixed

File: @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAltitudeHold.sqf [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+        _heli setVariable ["fza_ah64_altHoldActive", false, true];
+    };
+
+    //If the helicopters radar altitude is < 1428ft (435.25m) and current velocity is < 40kts
+    //then set the desired altitude to the current AGL altitude, otherwise set it to the
+    //current ASL altitude.
+    if (_curAltAGL < RAD_ALT_MAX_ALT && _curVel < ALT_HOLD_SPEED_SWITCH) then {
+        _heli setVariable ["fza_ah64_altHoldSubMode", "rad"];
+    } else {
+        _heli setVariable ["fza_ah64_altHoldSubMode", "bar"];
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAltitudeHold.sqf:37]: These are not public variables, whereas it is used as a public variable elsewhere. However, you may want to make sure it doesn't do a public variable message each time

**Review by mattysmith22 — CHANGES_REQUESTED** [review]

File: @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAltitudeHold.sqf (hunk 12 above)
Comment (BradMick) [reply @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAltitudeHold.sqf:37]: Fixed

File: @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAltitudeHold.sqf [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+    private _collRef_hi  = _collRef * 1.05;
+    if (fza_sfmplus_collectiveOutput >= _collRef_hi || fza_sfmplus_collectiveOutput <= _collRef_low) then {
+        _heli setVariable ["fza_ah64_altHoldActive", false, true];
+    };
+
+    //If the helicopters radar altitude is < 1428ft (435.25m) and current velocity is < 40kts
+    //then set the desired altitude to the current AGL altitude, otherwise set it to the
+    //current ASL altitude.
+    if (_curAltAGL < RAD_ALT_MAX_ALT && _curVel < ALT_HOLD_SPEED_SWITCH) then {
+        _heli setVariable ["fza_ah64_altHoldSubMode", "rad", true];
```
Comment (mattysmith22) [inline @AH-64D Apache Longbow/Addons/fza_ah64_sfmplus/functions/fmc/fn_fmcAltitudeHold.sqf:35]: This is going to setVariable on a public thing every single frame - very bad for performance. Let's only set if it isn't already the correct value

**Review by mattysmith22 — CHANGES_REQUESTED** [review]

**Review by mattysmith22 — APPROVED** [review]
Looks good. Can't check whether there are tabs in indentation from my mobile phone, so conditional on there being no tabs

**Comment by Ollieollieolllie** [comment]
no tabs present

**Pull request merged**